### ISSUE TITLE
Crash in `_mesa_ClearColor`

### ISSUE THREAD

**Issue by Ms2ger** [servo/servo#5822]
Ubuntu 12.04

```
(gdb) bt
#0  0x00007ffff190ba80 in _mesa_ClearColor () from /usr/lib/x86_64-linux-gnu/dri/libdricore.so
#1  0x00005555559d0ba1 in ffi::ClearColor::he88705059bfc8fb8TJb () at target/debug/build/gleam-d3837506b4549386/out/gl_bindings.rs:4533
#2  0x00005555559d0b0a in gl::clear_color::hc85255f2e50c601bUta () at /.cargo/git/checkouts/gleam-e81a762a5fab98bb/master/src/gl.rs:409
#3  0x00005555559caf82 in window::Window::new::ha334407313379a0booa () at /ports/glutin/window.rs:102
#4  0x0000555555aaa32b in create_window::h1cc1564068fa48bdGTa () at /ports/glutin/lib.rs:47
#5  0x00005555556f88f8 in main::h8dbf4b2c8d5229fbpaa () at main.rs:56
#6  0x00005555575d3919 in rust_try_inner ()
#7  0x00005555575d3906 in rust_try ()
#8  0x00005555575cb191 in rt::lang_start::h198f2d8a631cfb22yBJ ()
#9  0x00005555556f8d85 in main ()
```

**Comment by tdaede** [comment]
Ubuntu 12.04 has a very old version of Mesa (8.0.4, I believe). It might not even support creating an OpenGL 3.0 context for your hardware, which might be the cause of the crash.

**Comment by jdm** [comment]
I wonder if this is due to the gleam update in c82485874d1fe1db89a2dac61c6d612a75b8e1a1?

**Comment by Ms2ger** [comment]
It is.

**Comment by Ms2ger** [comment]
CC @glennw

**Comment by glennw** [comment]
The problem is caused here https://github.com/servo/servo/blob/master/components/canvas/Cargo.toml#L31.

This line links to libosmesa, which uses software rendering for the webgl implementation. What this means is that on Linux, both libosmesa and libgl are linked, which both define conflicting symbols. In your case on 12.04, the wrong function ends up getting called, which has a null context in mesa, thus the crash.

Linking to both osmesa and libgl works fine on ubuntu 14.04, 14.10 and arch - they don't run into the same problem as mesa on 12.04.

The documentation for osmesa is very sparse, but I did find a few references suggesting that linking to both libs was a bad idea, but these posts were also from several years ago.

The question is whether the fact that linking to both libs works in modern distros is by design, or a lucky coincidence. My brief scan of the mesa source code suggests it's more by good luck than anything else that it works on modern distros, but I could be mistaken there.

In any case, the end goal is to use hardware accelerated rendering for webgl, in which case this problem goes away. https://github.com/servo/servo/commit/79a5dae1702916c724e7c9e08ead44a1b90cdedb is the first step towards this. However, it still has the osmesa libs enabled as a fallback.

TL;DR: This problem should resolve itself once Emilio's work is complete and we don't use the osmesa context for webgl. In the short term, one option is to locally patch your servo - remove the line I linked to in Cargo.toml above, and change the resulting compile errors to panic!(). This means you can't run webgl tests, but would be better than the current situation you have.

There's other alternatives of course - but patching locally is one option to allow working on 12.04 while the hardware accelerated parts of webgl are being worked on.

**Comment by ecoal95** [comment]
I don't link explicitly except when I run tests, I think this shouldn't be the problem.

**Comment by glennw** [comment]
Yes, that's correct - I removed that line on a 12.04 VM and confirmed it fixes the problem :)
